src/runtime/pyscope.cs

```diff
@@ -0,0 +1,535 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Dynamic;
+
+namespace Python.Runtime
+{
+    public class PyScopeException : Exception
+    {
+        public PyScopeException(string message)
+            : base(message)
+        {
+
+        }
+    }
+
+    /// <summary>
+    /// Classes/methods have this attribute must be used with GIL obtained.
+    /// </summary>
+    public class PyGILAttribute : Attribute
+    {
+    }
+
+    [PyGIL]
+    public class PyScope : DynamicObject, IDisposable
+    {
+        public readonly string Name;
+
+        internal readonly IntPtr obj;
+
+        /// <summary>
+        /// the dict for local variables
+        /// </summary>
+        internal readonly IntPtr variables;
+
+        private bool isDisposed;
+
+        internal readonly PyScopeManager Manager;
+
+        public event Action<PyScope> OnDispose;
+        
+        public PyScope(IntPtr ptr, PyScopeManager manager)
+        {
+            if (Runtime.PyObject_Type(ptr) != Runtime.PyModuleType)
+            {
+                throw new PyScopeException("object is not a module");
+            }
+            if(manager == null)
+            {
+                manager = PyScopeManager.Global;
+            }
+            Manager = manager;
+            obj = ptr;
+            //Refcount of the variables not increase
+            variables = Runtime.PyModule_GetDict(obj);
+            if (variables == IntPtr.Zero)
+            {
+                throw new PythonException();
+            }
+            Runtime.PyDict_SetItemString(
+                variables, "__builtins__",
+                Runtime.PyEval_GetBuiltins()
+            );
+            this.Name = this.GetVariable<string>("__name__");
+        }
+
+        public PyDict Variables()
+        {
+            Runtime.XIncref(variables);
+            return new PyDict(variables);
+        }
+
+        public PyScope NewScope()
+        {
+            var scope = Manager.Create();
+            scope.ImportAll(this);
+            return scope;
+        }
+
+        public dynamic Import(string name, string asname = null)
+        {
+            Check();
+            if (asname == null)
+            {
+                asname = name;
+            }
+            var scope = Manager.TryGet(name);
```

Comment:
For `Try` the usual "protocol" is to let it return a `bool` and output to an `out` variable:

     PyScope scope;
     if (Manager.TryGet(name, out scope)) ...